app/src/main/java/packagename/app/com/appname/core/module/WebserviceModule.java

```diff
@@ -24,36 +25,45 @@
 
    public static final int CACHE_SIZE = 25 * 1024 * 1024;
 
-   /* TODO: Add here provides methods for retrofit interfaces. */
+   @Provides
+   @Singleton
+   Cache provideHttpCache(Context context) {
+      final File cacheDirectory = new File(context.getCacheDir()
+            .getAbsolutePath(), "HttpCache");
+      return new Cache(cacheDirectory, CACHE_SIZE);
+   }
 
    @Provides
    @Singleton
-   Picasso providePicasso(Context context) {
+   OkHttpClient provideHttpClient(Context context, Cache cache) {
+      final OkHttpClient client = new OkHttpClient();
+      client.setCache(cache);
+      return client;
+   }
+
+   @Provides
+   @Singleton
+   Picasso providePicasso(Context context, OkHttpClient client) {
       Picasso.Builder imageLoaderBuilder = new Picasso.Builder(context);
       imageLoaderBuilder.executor(Executors.newSingleThreadExecutor());
+      imageLoaderBuilder.downloader(new OkHttpDownloader(client));
+      imageLoaderBuilder.indicatorsEnabled(BuildConfig.IS_IDE_BUILD);
       return imageLoaderBuilder.build();
    }
 
    @Provides
    @Singleton
-   Retrofit provideRetrofit(Context context) {
+   Retrofit provideRetrofit(Context context, OkHttpClient client) {
       Retrofit.Builder builder =
             new Retrofit.Builder().baseUrl(context.getString(R.string.base_url))
                   .addConverterFactory(GsonConverterFactory.create());
       if (BuildConfig.DEBUG) {
-         OkHttpClient client = new OkHttpClient();
-         client.interceptors().add(new LoggingInterceptor());
+         client.interceptors()
+               .add(new LoggingInterceptor());
```

Comment:
should be replaced with:
         HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
         interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
from: com.squareup.okhttp:logging-interceptor:2.6.0